k9mail-library/src/main/java/com/fsck/k9/mail/ssl/DefaultTrustedSocketFactory.java

```diff
@@ -160,7 +169,18 @@ public Socket createSocket(Socket socket, String host, int port, String clientCe
         SSLSocketFactory socketFactory = sslContext.getSocketFactory();
         Socket trustedSocket;
         if (socket == null) {
-            trustedSocket = socketFactory.createSocket();
+            if (proxySettings.enabled) {
+                InetSocketAddress proxyAddress = new InetSocketAddress(proxySettings.host, proxySettings.port);
+                Proxy proxy = new Proxy(Type.SOCKS, proxyAddress);
+
+                Socket underlying = new Socket(proxy);
+                InetSocketAddress serverAddress = new InetSocketAddress(host, port);
+                underlying.connect(serverAddress, SOCKET_CONNECT_TIMEOUT);
+
+                trustedSocket = socketFactory.createSocket(underlying, proxySettings.host, proxySettings.port, true);
```

Comment:
This feels wrong:

You create a plain Socket with a proxy.
You connect that plain Socket to the end point.
Then you create (and connect, but this is hidden by the SSL API) that SSL socket wrapped round that socket to the proxy.

Don't you now have a connection to the proxy via the server via the proxy?

i.e

`Client -> Proxy -> Server -> Proxy`

Reply:
I think we want:

```
                InetSocketAddress proxyAddress = new InetSocketAddress(proxySettings.host, proxySettings.port);
                Proxy proxy = new Proxy(Type.SOCKS, proxyAddress);
                Socket underlying = new Socket(proxy);
                InetSocketAddress serverAddress = new InetSocketAddress(host, port);
                underlying.connect(serverAddress, SOCKET_CONNECT_TIMEOUT);
                trustedSocket = socketFactory.createSocket(underlying, host, port, true);
```

That is, connect a socket to SOCKS PROXY.
Use that underlying socket to connect to the remote socket.
Upgrade the remote socket proxied connection to SSL.